### ISSUE TITLE
linter: `"env"` config doesn't block use of disallowed globals

### ISSUE THREAD

**Issue by rschristian** [oxc-project/oxc#16864]
### What version of Oxlint are you using?

1.32.0

### What command did you run?

oxlint src/index.js

### What does your `.oxlintrc.json` config file look like?

```jsonc
{
	"env": {
		"browser": false,
		"node": false,
	}
}
```

### What happened?

Despite disabling both browser & Node envs, oxlint seemingly has no issue with any of the following lines, despite relying upon browser & Node globals:

```js
document.createElement('div');
process.exit(1);
```

Along those lines, I can't actually spot any effect from the `"env"` setting. It's [documented](https://oxc.rs/docs/guide/usage/linter/config-file-reference.html#env), so I'd hope it was hooked up, but it doesn't warn on disallowed globals, it doesn't warn on disallowed built-in access, nor does it warn upon disallowed syntax.

Not quite sure what it's meant to be doing.

**Comment by connorshea** [comment]
I think it may be setting all the globals under that Eng to "false" which just means it's readonly, not disallowed.

**Comment by rschristian** [comment]
Sorry, so telling `oxlint` that we're not in a browser env (`"env": { "browser": false }}`) only blocks me from overriding browser globals? Why would I be overriding globals for an environment I'm not using?

[The docs also seem to indicate otherwise](https://oxc.rs/docs/guide/usage/linter/config-file-reference.html#env):

> Environments specify what global variables are predefined.

If `"browser": false` is set, then the browser global vars should not be defined, no?

**Comment by connorshea** [comment]
Apologies, yeah you're correct that they should not be defined in that case: https://github.com/oxc-project/oxc/blob/main/crates/oxc_linter/src/config/env.rs#L61-L70

**Comment by tt-a1i** [comment]
Looked into this - the `env` config is actually working correctly, but there's a missing piece:

**How it works:**
- `env` only **declares** which global variables exist - it doesn't **enforce** anything on its own
- Setting `"browser": false` means `document`, `window`, etc. won't be recognized as globals
- But to actually **report errors** when undefined globals are used, you need the `no-undef` rule

**The catch:** `eslint/no-undef` is in the `nursery` category, so it's **not enabled by default**.

**Solution:**
```json
{
  "env": {
    "browser": false,
    "node": false
  },
  "rules": {
    "eslint/no-undef": "error"
  }
}
```

This will report `document` and `process` as undefined.

The test cases in `no_undef.rs` confirm this behavior:
```rust
// passes because browser env is enabled
("URLSearchParams;", None, Some(json!({"env": { "browser": true }})))

// fails because window is not in default env (builtin only)
("window;", None, None)
```

Perhaps the docs could clarify that `env` works in conjunction with `no-undef` to catch disallowed globals?

**Comment by rschristian** [comment]
Sorry for the super late reply, holidays have had me super busy.

> But to actually **report errors** when undefined globals are used, you need the `no-undef` rule

Hrm, that's an interesting approach. Not sure I fully understand the value in setting `env` values if `no-undef` doesn't implicitly get activated alongside it.

> Perhaps the docs could clarify that `env` works in conjunction with `no-undef` to catch disallowed globals?

That would indeed help a lot, at least it would've for me.

---

I'm going to close this out as my question has been answered as I've misunderstood. Adding docs or not is up to you of course, maybe I'm the odd one out here.

Thanks for taking the time to answer though! I really appreciate it
